### PR TITLE
fix(app): ignore stale assistant messages for running state

### LINKED ISSUE #95: [Bug] Completed sessions can stay marked as running after stale incomplete assistant messages
## What happened?

A completed PawWork session can continue to appear as running when its history contains an older assistant message without `time.completed`. In the observed `Greeting` session, the top session progress indicator stayed visible above the conversation title and the sidebar item kept showing the running spinner even though later assistant turns had completed and the latest assistant message ended with `finish=stop`.

Local inspection of the backing OpenCode SQLite data found one stale assistant message, `msg_dafe091a7001X3Oa6sn8pBq7lc`, with no `time.completed`, no finish value, and only `step-start` plus an empty `reasoning` part. The same session later contains multiple completed assistant messages, including a final assistant message with `finish=stop` and a completed timestamp. Two other nearby sessions in the same app state did not have incomplete assistant messages and did not show the same false-running evidence.

The current UI checks are broad enough that one stale historical assistant message can keep the whole session marked busy. Relevant reads in the current code are `packages/app/src/pages/session/message-timeline.tsx`, `packages/app/src/pages/layout/sidebar-items.tsx`, and `packages/app/src/pages/session.tsx`, where running/busy state is derived from any assistant message lacking `time.completed` in addition to `session_status`.

## Steps to reproduce

1. Open PawWork v0.2.4 on macOS.
2. Open a session whose history contains an older assistant message without `time.completed`, followed by later completed assistant turns.
3. Observe the session list and session header.

Observed concrete local case: session title `Greeting`, session id `ses_2544c191effeu1Nw5UECHonKbl`, database `~/.local/share/opencode/opencode.db`. The stale message was created at local time 2026-04-21 19:50:23 and the latest completed assistant message was created at local time 2026-04-21 20:09:16.

## What did you expect to happen?

Once the latest run has stopped and the current `session_status` is idle, old incomplete assistant records should not keep the entire session marked as running forever. The top progress indicator, title spinner, sidebar spinner, and busy/follow-up logic should reflect the current turn state, not stale historical corruption.

## Environment

PawWork version: v0.2.4

OS version: macOS 26.3.1 (a), build 25D771280a

Can reproduce again: Sometimes. The false-running symptom was observed in the active app session. Clicking back into the affected `Greeting` session later triggered the already tracked renderer crash in #94, so this issue should be treated as the false-running state bug while #94 tracks the crash path.

## Extra context

This is separate from #94. During follow-up verification, clicking the same affected session reproduced the #94 renderer crash (`TypeError: Cannot read properties of undefined (reading '0')`), but the root evidence for this issue is the stale incomplete assistant message causing false current busy state before the crash path is hit.

### PR DESCRIPTION
**Pull request by Astro-Han**
## Summary

- Add a shared `isSessionRunning` helper for app session running-state checks.
- Treat live non-idle `session_status` as authoritative, and only fall back to message history when the latest message is an incomplete assistant message.
- Replace the broad historical incomplete-assistant scans in the session timeline, sidebar item, and session busy helper.

## Why

A stale incomplete assistant message from an older turn can remain in session history even after later turns complete. The previous UI checks treated any historical incomplete assistant message as current work, which kept completed sessions showing progress/spinners.

## Related Issue

Fixes #95

## How To Verify

```bash
cd packages/app && bun run test:unit src/pages/session/session-running-state.test.ts
cd packages/app && bun run test:unit
cd packages/app && bun run typecheck
bun turbo typecheck
bun turbo test:ci
```

## Screenshots or Recordings

Not included. This is a state-derivation fix covered by unit tests, with no intentional visual design change.

## Checklist

- [x] I ran the relevant verification steps
- [x] I tested visible changes manually when needed
- [x] I am targeting the `dev` branch

## Summary by CodeRabbit

* **Refactor**
  * Consolidated session state determination logic into a centralized shared helper function, reducing code duplication and improving maintainability across multiple application components.

* **Tests**
  * Added comprehensive test coverage for session running state evaluation, including edge cases and various status conditions.

### CODE REVIEW

File: packages/app/src/pages/session/session-running-state.ts [hunk 1]
```diff
@@ -0,0 +1,10 @@
+import type { Message, SessionStatus } from "@opencode-ai/sdk/v2/client"
+
+const idle = { type: "idle" as const }
+
+export function isSessionRunning(status: SessionStatus | undefined, messages: readonly Message[] | undefined): boolean {
+  if ((status ?? idle).type !== "idle") return true
+
+  const latest = messages?.at(-1)
+  return latest?.role === "assistant" && typeof latest.time.completed !== "number"
```
[the remark this answers is not in the document]
Comment (Astro-Han) [reply packages/app/src/pages/session/session-running-state.ts]: Fixed in `8ca5931b1`. I also rechecked the SDK types: the role guard does narrow to `AssistantMessage`, but the optional access is harmless and makes this helper more defensive against malformed message payloads.

File: packages/app/src/pages/layout/sidebar-items.tsx [hunk 2]
```diff
@@ -170,18 +171,7 @@ export const SessionItem = (props: SessionItemProps): JSX.Element => {
   })
   const isWorking = createMemo(() => {
     if (hasPermissions()) return false
```
Comment (Astro-Han) [inline packages/app/src/pages/layout/sidebar-items.tsx:173]: 旧逻辑用 `findLast` 扫描整段历史消息来寻找未完成的 assistant，替换后语义更清晰：先以 `session_status` 为权威，idle 时才看最新消息。消除了 stale-message 导致的假阳性。

File: packages/app/src/pages/session.tsx [hunk 3]
```diff
@@ -1606,12 +1607,7 @@ export default function Page() {
       return out
```
Comment (Astro-Han) [inline packages/app/src/pages/session.tsx:1607]: 注意语义收窄：旧逻辑用 `.some()` 扫描所有消息，而 `isSessionRunning` 只看最后一条。这是修复 stale message 的核心，但值得确认没有调用方依赖“任意历史未完成即 busy”的行为。

File: packages/app/src/pages/session/message-timeline.tsx [hunk 4]
```diff
@@ -261,8 +262,7 @@ export function MessageTimeline(props: {
     return sync.data.session_status[id] ?? idle
   })
   const working = createMemo(() => {
-    const status = sessionStatus() ?? idle
-    return !!pending() || status.type !== "idle"
+    return isSessionRunning(sessionStatus(), sessionMessages())
```
Comment (Astro-Han) [inline packages/app/src/pages/session/message-timeline.tsx:265]: 替换正确。`pending()` memo 仍被保留用于下游 `activeMessageID`（第 285 行），职责已从“判断 running”收窄为“定位当前生成中的消息”，分工清晰。

File: packages/app/src/pages/session/session-running-state.ts [hunk 5]
```diff
@@ -0,0 +1,10 @@
+import type { Message, SessionStatus } from "@opencode-ai/sdk/v2/client"
+
+const idle = { type: "idle" as const }
+
+export function isSessionRunning(status: SessionStatus | undefined, messages: readonly Message[] | undefined): boolean {
+  if ((status ?? idle).type !== "idle") return true
+
```
Comment (Astro-Han) [inline packages/app/src/pages/session/session-running-state.ts:8]: 逻辑简洁。建议加一行 JSDoc 说明设计意图：当 status 为 idle 时仅检查最新消息，避免后人误以为需要扫描历史。

File: packages/app/src/pages/session/session-running-state.test.ts [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+      assistant("msg_stale", 2),
+      user("msg_user_2", 3),
+      assistant("msg_done", 4, 5, "stop"),
+    ]
+
+    expect(isSessionRunning(idle, messages)).toBe(false)
+  })
+
+  test("ignores a stale incomplete assistant message when a later user message exists", () => {
+    const messages = [user("msg_user_1", 1), assistant("msg_stale", 2), user("msg_user_2", 3)]
```
Comment (Astro-Han) [inline packages/app/src/pages/session/session-running-state.test.ts:55]: stale-message 场景覆盖很好。建议再补一个边界 case：`messages = []` + idle status 应返回 false，避免空数组回归。

**Review by Astro-Han — COMMENTED** [review]
总体逻辑正确，测试覆盖充分。几点 inline 供参考（无阻塞问题）。

**Review by Astro-Han — COMMENTED** [review]
Overall: Clean fix with good test coverage. One edge case suggestion.

## Core Logic Review

The new `isSessionRunning` helper consolidates duplicate logic across 3 files — good refactor. The logic is:

```ts
if ((status ?? idle).type !== "idle") return true
const latest = messages?.at(-1)
return latest?.role === "assistant" && typeof latest.time?.completed !== "number"
```

**Key insight**: Trust live `session_status` over message history. Only fallback to checking the *latest* message, not scanning all history. This correctly fixes the stale-incomplete-assistant bug.

## Edge Case: Empty Message Array

When `messages` is an empty array `[]`:
- `messages?.at(-1)` returns `undefined`
- `undefined?.role === "assistant"` is `false`
- Returns `false` ✓ (idle state)

This is correct behavior, but worth adding a test case for clarity.

## Minor Nits

1. **Test file import path**: `import { isSessionRunning } from "./session-running-state"` — works, but explicit extension would be clearer for future tooling.

2. **Test helper naming**: The `user` and `assistant` factory functions are minimal. Consider adding JSDoc or renaming to `makeUserMessage`/`makeAssistantMessage` for discoverability.

These are nits; not blocking.

## Suggestion: Add Empty Array Test

```ts
test("returns false when messages array is empty", () => {
  expect(isSessionRunning(idle, [])).toBe(false)
})
```

This documents the edge case and ensures future refactorers don't accidentally change behavior.

File: packages/app/src/pages/layout/sidebar-items.tsx (hunk 2 above)
Comment (Astro-Han) [reply packages/app/src/pages/layout/sidebar-items.tsx:173]: Confirmed. This is the intended sidebar behavior: live `session_status` remains authoritative, and idle sessions only fall back to the latest message so stale incomplete assistant history cannot keep the row busy.

File: packages/app/src/pages/session.tsx (hunk 3 above)
Comment (Astro-Han) [reply packages/app/src/pages/session.tsx:1607]: Confirmed. I rechecked the three call sites and did not find a legitimate dependency on historical incomplete assistants meaning busy. The only remaining historical lookup is `pending()` in `message-timeline.tsx`, and that is scoped to locating the active assistant message, not deriving session running state. The stale assistant followed by completed assistant and stale assistant followed by user cases are covered in `session-running-state.test.ts`.

File: packages/app/src/pages/session/message-timeline.tsx (hunk 4 above)
Comment (Astro-Han) [reply packages/app/src/pages/session/message-timeline.tsx:265]: Confirmed. `pending()` is intentionally kept for active-message anchoring only; running-state derivation now goes through `isSessionRunning`.

File: packages/app/src/pages/session/session-running-state.ts (hunk 5 above)
Comment (Astro-Han) [reply packages/app/src/pages/session/session-running-state.ts:8]: Fixed in `de0bffb16` with a one-line JSDoc describing the status-first, latest-message fallback rule.

File: packages/app/src/pages/session/session-running-state.test.ts (hunk 6 above)
Comment (Astro-Han) [reply packages/app/src/pages/session/session-running-state.test.ts:55]: Fixed in `de0bffb16` by adding the idle plus empty messages boundary test.

**Comment by Astro-Han** [comment]
@coderabbitai review

File: packages/app/src/pages/session/session-running-state.test.ts [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+    expect(isSessionRunning(idle, messages)).toBe(true)
+  })
+
+  test("returns false when there are no messages and status is idle", () => {
+    expect(isSessionRunning(idle, [])).toBe(false)
+  })
+
+  test("returns false when there are no assistant messages and status is idle", () => {
+    expect(isSessionRunning(idle, [user("msg_user", 1)])).toBe(false)
+  })
```
[the remark this answers is not in the document]
Comment (Astro-Han) [reply packages/app/src/pages/session/session-running-state.test.ts:80]: Fixed in `7b4497917`. Added explicit coverage for `undefined` status and messages, including empty messages, user-only messages, and the latest incomplete assistant fallback.
